multi_pong/src/network/mod.rs

```diff
@@ -52,221 +45,76 @@ impl NetworkCommunication {
-                serialized.send(Packet::Ping(id)).await.unwrap();
-            }
-            let delay = delay_for(Duration::new(0, 10_000_000)).fuse();
-            let fg_to = from_foreground.next().fuse();
-            let to_fg = deserialized.next().fuse();
-            pin_mut!(delay, fg_to, to_fg);
-            select_biased! {
-                _ = delay => (),
-                inst = fg_to => {
-                    if let Some(inst) = inst {
-                        match inst {
-                            Instruction::Disconnect(_) => break,
-                            Instruction::SendPacket(packet) => {
-                                serialized.send(packet).await.unwrap();
-                            },
-                        }
-                    }
-                },
-                msg = to_fg => {
-                    if let Some(msg) = msg {
-                        match msg.unwrap() {
-                            Packet::Ping(packet_id) => serialized.send(Packet::Pong(packet_id)).await.unwrap(),
-                            Packet::Pong(packet_id) => if packet_id == id {
-                                println!("Ping: {}", now.elapsed().as_millis());
-                                pong_received = true;
-                            },
-                            others => to_foreground.send(ResponseState::PacketReceived(others)).await.unwrap()
-                        }
-                    }
-                },
-            };
+            ping_send
+                .unbounded_send(Packet::Ping(GLOBAL_TIMER.elapsed().as_millis()))
+                .unwrap();
+            delay_for(Duration::new(0, 100_000_000)).await;
         }
-    };
-    client.await
+    });
+    NetworkCommunication::new(recv, send, Side::Server)
 }
 
 #[tokio::main]
-async fn server_network_loop(mut from_foreground: UnboundedReceiver<Instruction>, mut to_foreground: UnboundedSender<ResponseState>) {
-    let mut listener = TcpListener::bind("0.0.0.0:4001").await.unwrap();
-    let (tx, rx) = async_std::sync::channel(1);
-    let server = async move {
-        let mut incoming = listener.incoming();
-        while let Some(conn) = incoming.next().await {
-            match conn {
-                Err(e) => println!("connection failed: {:?}", e),
-                Ok(sock) => {
-                    if let Err(e) = tx.try_send(sock) {
-                        match e {
-                            TrySendError::Full(_sock) => {
-                                //notify the connection
-                            },
-                            TrySendError::Disconnected(_sock) => {
-                                //notify the connection
-                                break;
-                            }
-                        }
-                    }
-                }
-            }
-        }
-    };
+pub async fn create_client_background_loop(addr: &str) -> NetworkCommunication {
+    let port: u16 = rand::thread_rng().gen();
```

Comment:
https://rust-lang-nursery.github.io/rust-cookbook/net/server.html#listen-on-unused-port-tcpip

SocketAddrV4 assigns a random port when setting port to 0.